### PR TITLE
Master timesheet days jem

### PR DESCRIPTION
**Pull request by jem-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 1]
```diff
@@ -65,6 +72,65 @@ def write(self, values):
         self.filtered(lambda t: t.project_id)._timesheet_postprocess(values)
         return result
 
+    @api.model
+    def fields_view_get(self, view_id=None, view_type='form', toolbar=False, submenu=False):
+        """ Set the correct widget for `unit_amount` field on tree and form views, depending on company UoM configuration (`float_toggle` for Days, `float_time` for Hours, ...) """
+        result = super(AccountAnalyticLine, self).fields_view_get(view_id=view_id, view_type=view_type, toolbar=toolbar, submenu=submenu)
```
Comment (antonylesuisse) [inline addons/hr_timesheet/models/hr_timesheet.py]: If can find attribute company_time_unit, replace that attribute by the relevant wiget.


                    <field name="unit_amount" string="Time (Hours)" sum="Total time" company_time_unit="1"/>

File: addons/web/static/src/js/fields/basic_fields.js [hunk 2]
```diff
@@ -883,6 +883,119 @@ var FieldFloatTime = FieldFloat.extend({
...
+        if (this.mode === 'edit') {
+            this.tagName = 'button';
+        }
+
+        if (this.attrs.digits) {
+            this.nodeOptions.digits = JSON.parse(this.attrs.digits);
+        }
+
+        this._range = this.nodeOptions.range || [0.0, 0.5, 1.0];
+        this._factor = this.nodeOptions.factor || 1.0;
```
Comment (antonylesuisse) [inline addons/web/static/src/js/fields/basic_fields.js]: factor ==  session_info.timesshet_factor

File: addons/web/static/tests/fields/basic_fields_tests.js [hunk 3]
```diff
@@ -3264,6 +3264,90 @@ QUnit.module('basic_fields', {
...
+
+    QUnit.test('float_factor field in form view', function (assert) {
+        assert.expect(4);
+
+        var form = createView({
+            View: FormView,
+            model: 'partner',
+            data: this.data,
+            arch:'<form string="Partners">' +
+                    '<sheet>' +
```
Comment (ged-odoo) [inline addons/web/static/tests/fields/basic_fields_tests.js]: no need for sheet tag

File: addons/web/static/src/js/fields/field_utils.js [hunk 4]
```diff
@@ -173,6 +173,21 @@ function formatFloat(value, field, options) {
     return formatted.join(l10n.decimal_point);
 }
 
+
+/**
+ * Returns a string representing a float value, from a float converted with a
+ * factor.
+ *
+ * @param {float} value
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/field_utils.js]: number

File: addons/web/static/src/js/fields/field_utils.js [hunk 5]
```diff
@@ -528,6 +543,18 @@ function parseMonetary(value, field, options) {
     return parseFloat(values[0] === currency.symbol ? values[1] : values[0]);
 }
 
+/**
+ * Parse a String containing float and unconvert it with a conversion factor
+ *
+ * @param {float} [options.factor]
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/field_utils.js]: number. also, document its default value

File: addons/web/static/src/js/fields/field_utils.js [hunk 6]
```diff
@@ -173,6 +173,21 @@ function formatFloat(value, field, options) {
     return formatted.join(l10n.decimal_point);
 }
 
+
+/**
+ * Returns a string representing a float value, from a float converted with a
+ * factor.
+ *
+ * @param {float} value
+ * @param {float} [options.factor]
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/field_utils.js]: document default value

File: addons/web/static/src/js/fields/basic_fields.js [hunk 7]
```diff
@@ -898,6 +898,140 @@ var FieldFloatTime = FieldFloat.extend({
...
+
+    /**
+     * Clicking on the button triggers the change of value; the next one of
+     * the range will be displayed.
+     *
+     * @private
+     * @param {OdooEvent} ev
+     */
+    _onClick: function(ev) {
+        if (this.mode == 'edit') {
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/basic_fields.js]: ===

File: addons/web/static/src/js/fields/basic_fields.js [hunk 8]
```diff
@@ -898,6 +898,140 @@ var FieldFloatTime = FieldFloat.extend({
...
+     * one is not in the range, the next value of the closest one will be chosen.
+     *
+     * @private
+     * @returns float: the next formatted value in the range
+     */
+    _nextValue: function () {
+        var range = this.nodeOptions.range;
+        var val =  utils.closestNumber(this._getDisplayedValue(), range);
+        var index = _.indexOf(range, val);
+        if (index != -1) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/basic_fields.js]: !==

File: addons/web/static/src/js/fields/basic_fields.js [hunk 9]
```diff
@@ -898,6 +898,140 @@ var FieldFloatTime = FieldFloat.extend({
...
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Get the display value but in real type to use it in calculations
+     *
+     * @private
+     * @returns float: the current formatted value
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/basic_fields.js]: invalid jsdoc

File: addons/web/static/src/js/fields/abstract_field.js [hunk 10]
```diff
@@ -168,9 +168,12 @@ var AbstractField = Widget.extend({
         // the 'widget' attrs if is is given, and if it is a valid key, with a
         // fallback on the field type, ensuring that the value is formatted and
         // displayed according to the choosen widget, if any.
-        this.formatType = this.attrs.widget in field_utils.format ?
-                            this.attrs.widget :
-                            this.field.type;
+        if (!this.formatType) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/fields/abstract_field.js]: why is this needed?

File: addons/hr_timesheet/static/src/js/timesheet_uom.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+var TimesheetToggle = basic_fields.FieldFloatToggle.extend({
+    formatType: 'float_factor',
+    init: function(parent, name, record, options) {
+        options = options || {};
+        var fieldsInfo = record.fieldsInfo[options.viewType || 'default'];
+        var attrs = options.attrs || (fieldsInfo && fieldsInfo[name]) || {};
+
+        var has_range = _.contains(_.keys(attrs.options || {}), 'range');
+
+        this._super(parent, name, record, options);
```
[the remark this answers is not in the document]
Comment (jem-odoo) [reply addons/hr_timesheet/static/src/js/timesheet_uom.js]: if there was no range defined on the node, the widget (super) is gonna set its default. So after the call to super you can not know if we can force or not the rage option, as we want to give prior to the option defined on the node itself.

File: addons/web/static/src/js/fields/abstract_field.js (hunk 10 above)
Comment (jem-odoo) [reply addons/web/static/src/js/fields/abstract_field.js]: float_toggle widget have the float_factor formatter. So there is no float_toggle formatter. With the code before, the formatter will be the field type one, as the name of the formatter is not the same as the widget in the registry. 
This commit gives priority to the formatter defined on the widget and fallback on a deduction based on formatter/widget name.

Comment (ged-odoo) [reply addons/web/static/src/js/fields/abstract_field.js]: usually we simply call super in the sub widget, and then override the specific key after (in this case, formatType)
